modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java

```diff
@@ -0,0 +1,391 @@
+            @Override
+            public CompletableFuture<Void> onUpdate(WatchEvent event) {
+                return handleTriggerKeyUpdate(event);
+            }
+
+            @Override
+            public void onError(Throwable e) {
+                // No-op.
+            }
+        };
+
+        messagingService.addMessageHandler(TableMessageGroup.class, this::handleMessage);
+    }
+
+    @Override
+    public CompletableFuture<Void> start() {
+        metaStorageManager.registerExactWatch(RECOVERY_TRIGGER_KEY, watchListener);
+
+        return nullCompletedFuture();
+    }
+
+    @Override
+    public void stop() throws Exception {
+        metaStorageManager.unregisterWatch(watchListener);
+    }
+
+    /**
+     * Updates assignments of the table in a forced manner, allowing for the recovery of raft group with lost majorities. It is achieved via
+     * triggering a new rebalance with {@code force} flag enabled in {@link Assignments} for partitions where it's required. New pending
+     * assignments with {@code force} flag remove old stable nodes from the distribution, and force new Raft configuration via "resetPeers"
+     * so that a new leader could be elected.
+     *
+     * @param zoneId Distribution zone ID.
+     * @param tableId Table ID.
+     * @return Operation future.
+     */
+    public CompletableFuture<Void> resetPartitions(int zoneId, int tableId) {
+        return processNewRequest(new ManualGroupUpdateRequest(UUID.randomUUID(), zoneId, tableId));
+    }
+
+    /**
+     * Returns partition states for all zones' partitions in the cluster. Result is a mapping of {@link TablePartitionId} to the mapping
+     * between a node name and a partition state.
+     *
+     * @param zoneName Zone name.
+     * @return Future with the mapping.
+     */
+    public CompletableFuture<Map<TablePartitionId, Map<String, LocalPartitionState>>> partitionStates(String zoneName) {
+        int latestCatalogVersion = catalogManager.latestCatalogVersion();
+        Optional<CatalogZoneDescriptor> zoneDesciptorOptional = catalogManager.zones(latestCatalogVersion).stream()
+                .filter(catalogZoneDescriptor -> catalogZoneDescriptor.name().equals(zoneName))
+                .findAny();
+
+        if (zoneDesciptorOptional.isEmpty()) {
+            return CompletableFuture.failedFuture(new DistributionZoneNotFoundException(zoneName, null));
+        }
+
+        CatalogZoneDescriptor zoneDescriptor = zoneDesciptorOptional.get();
+
+        Set<NodeWithAttributes> logicalTopology = dzManager.logicalTopology();
+
+        LocalPartitionStatesRequest localPartitionStatesRequest = MSG_FACTORY.localPartitionStatesRequest()
+                .zoneId(zoneDescriptor.id())
+                .catalogVersion(latestCatalogVersion)
+                .build();
+
+        Map<TablePartitionId, Map<String, LocalPartitionState>> result = new ConcurrentHashMap<>();
+        List<CompletableFuture<?>> futures = new ArrayList<>();
```

Comment:
It would be possible to use an array right away, Meow.

Reply:
Why bother. Do you want me to re-write the code with array?

Reply:
It would be ameowazing.